adapters/smarthub/smarthub.go

```diff
@@ -0,0 +1,187 @@
+package smarthub
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"text/template"
+
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const (
+	ADAPTER_VER = "1.0.0"
+)
+
+type adapter struct {
+	endpoint template.Template
+}
+
+type bidExt struct {
+	MediaType string `json:"mediaType"`
+}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	template, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+
+	bidder := &adapter{
+		endpoint: *template,
+	}
+
+	return bidder, nil
+}
+
+func (a *adapter) getImpressionExt(imp *openrtb2.Imp) (*openrtb_ext.ExtSmartHub, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Bidder extension not provided or can't be unmarshalled",
+		}
+	}
+
+	var smarthubExt openrtb_ext.ExtSmartHub
+	if err := json.Unmarshal(bidderExt.Bidder, &smarthubExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: "Error while unmarshaling bidder extension",
+		}
+	}
+
+	return &smarthubExt, nil
+}
+
+func (a *adapter) buildEndpointURL(params *openrtb_ext.ExtSmartHub) (string, error) {
+	endpointParams := macros.EndpointTemplateParams{
+		Host:      params.Host,
+		AccountID: params.Seat,
+		SourceId:  params.Token,
```

Comment:
I see you pass all three parameters to your URL. Do you need all three to NOT be empty? As of right now, you marked them as `required` inside `static/bidder-params/smarthub.json` but they can still be zero-lenght strings. In case you need that any or all three of your parameters to not be empty you could add the `minLength` property like:
```
   {
     "$schema": "http://json-schema.org/draft-04/schema#",
     "title": "SmartHub Adapter Params",
     "description": "A schema which validates params accepted by the SmartHub adapter",
     "type": "object",
     "properties": {
       "host": {
         "type": "string",
         "description": "Network host to send request"
 +       "minLength": 1
       },
       "seat": {
         "type": "string",
         "description": "Seat"
 +       "minLength": 1
       },
       "token": {
         "type": "string",
         "description": "Token"
 +       "minLength": 1
       }
     },
     "required": [
       "host",
       "seat",
       "token"
     ]
   }
static/bidder-params/smarthub.json
```
Without the additions shown above, an all empty parameter `req.Imp[i].Ext` would be allowed. In other words, for the moment, a test case like this is currently passing:
```
   var invalidParams = []string{
   	``,
   	`null`,
   	`true`,
   	`5`,
   	`[]`,
   	`{}`,
   	`{"anyparam": "anyvalue"}`,
   	`{"host":"smarthub.test"}`,
   	`{"seat":"9Q20EdGxzgWdfPYShScl"}`,
   	`{"token":"Y9Evrh40ejsrCR4EtidUt1cSxhJsz8X1"}`,
   	`{"seat":"9Q20EdGxzgWdfPYShScl", "token":"alNYtemWggraDVbhJrsOs9pXc3Eld32E"}`,
   	`{"host":"smarthub.test", "token":"LNywdP2ebX5iETF8gvBeEoB6Cam64eeq"}`,
   	`{"host":"smarthub.test", "seat":"9Q20EdGxzgWdfPYShScl"}`,
 + 	`{"host":"", "seat":"", "token":""}`,  // <-- all empty fields test case is passing right now
   }
adapters/smarthub/params_test.go
```

Reply:
If that's what you intended, it's fine

Reply:
Yes, thanks a lot for pointing the mistake! All three parameters mustn't be an empty string. I've fixed this in `static/bidder-params/smarthub.json` and also added this cases to test into `invalidParams`